### PR TITLE
Cleaned up the config files, code is ready for review.

### PR DESCRIPTION
**Pull request by YannickMeeus**
Have at it.

Tests have been copied across from the MongoDB driver, in order to actually run them you'll have to set up a test Azure DocumentDB account and update the credentials in the test configuration file.

### CODE REVIEW

File: .nuget/NEventStore.nuspec [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
```
Comment (PeterStephenson) [inline .nuget/NEventStore.nuspec]: Is this file needed?

File: .nuget/packages.config [hunk 2]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="RavenDB.Server" version="2.0.2360" />
```
Comment (PeterStephenson) [inline .nuget/packages.config]: We using Raven etc? Or is this "inherited" from NES?

Comment (YannickMeeus) [inline .nuget/packages.config]: Probably inherited, yeah. I think it was required for the global test suite, which we're not using, I'll go through those later and trim the config file down to required references.

File: .nuget/NEventStore.nuspec (hunk 1 above)
Comment (YannickMeeus) [inline .nuget/NEventStore.nuspec]: File removed

**Comment by YannickMeeus** [comment]
Quick comment for those interested, the persistence component has been coded against the in-house coding standards and rules, so they've diverged from the NEventStore coding standards.

File: NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                    throw new Error('Error querying for documents with Logical_Key_Constraint_Failures:' + err.message);
+                }
+
+                if (results.length > 0)
+                {
+                    throw new Error('Commit failed due to Logical_Key_Constraint_Failure');
+                }
+                else if(responseOptions.continuation){
+                    HasDuplicates(responseOptions.continuation);
+                }
```
Comment (YannickMeeus) [inline NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js]: Seems to be not applicable, will dig more into the sample I've based this on.

File: NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+﻿function LogicalKey_Constraint() {
+    var context = getContext();
+    var collection = context.getCollection();
+    var request = context.getRequest();
+    var docToUpdate = request.getBody();
+
+    HasDuplicates();
+
+    function HasDuplicates(continuation)
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js]: is the continuation stuff needed? As we have no need to page (we'll error as soon as the first result is found)

File: NEventstore.Persistence.DocumentDB/DocumentCommit.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        [JsonProperty(TypeNameHandling = TypeNameHandling.All)]
+        public IList<string> Payload { get; set; }
+
+        public bool Dispatched { get; set; }
+
+        public long CheckpointNumber { get; set; }
+    }
+
+    internal class DateEpoch
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentCommit.cs]: Separate file?

Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentCommit.cs]: Also could you describe why this is needed please? :)

File: NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        private readonly Func<long> getLastCheckPointNumber;
+        private readonly Func<LongCheckpoint> getNextCheckPointNumber;
+        private readonly LongCheckpoint checkPointZero;
+
+        public DocumentPersistenceEngine(DocumentClient client, DocumentPersistenceOptions options)
+        {
+            this.Client = client;
+            this.Options = options;
+
+            this.getLastCheckPointNumber = () => this.TryExecute(() =>
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs:30]: Is this thread safe?
As discussed: Compared to the mongo one (similar implementation)
https://github.com/NEventStore/NEventStore.Persistence.MongoDB/blob/master/src/NEventStore.Persistence.MongoDB/MongoPersistenceEngine.cs

The Mongo one catches unique constraint violations on the checkpoint number index to "solve" the thread safety issue

File: NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
+﻿function LogicalKey_Constraint() {
+    var context = getContext();
+    var collection = context.getCollection();
+    var request = context.getRequest();
+    var documentToUpdate = request.getBody();
+
+    HasDuplicates();
+
+    function HasDuplicates(continuation)
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js]: is continuation needed here?

File: NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+
+        public void Initialize()
+        {
+            this.EnsureDatabaseExists();
+            var collections = this.Client.ReadDocumentCollectionFeedAsync(UriFactory.CreateDatabaseUri(this.Options.DatabaseName)).GetAwaiter().GetResult();
+            this.EnsureCollectionExists(collections, this.Options.CommitCollectionName);
+            this.EnsureCollectionExists(collections, this.Options.StreamHeadCollectionName);
+            this.EnsureCollectionExists(collections, this.Options.SnapshotCollectionName);
+
+            var logicalKeyConstraint = new Trigger
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs]: Can you check we have all the same unique indexes as mongo please?

File: NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+                return;
+            }
+
+            documentToUpdate.Dispatched = true;
+            this.Client.ReplaceDocumentAsync(UriFactory.CreateDocumentUri(this.Options.DatabaseName, this.Options.CommitCollectionName, documentToUpdate.Id), documentToUpdate).GetAwaiter().GetResult();
+        }
+
+        public void Purge(string bucketId)
+        {
+            Logger.Warn(Messages.PurgingStorage);
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs]: Should this have an implementation? :P

File: NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+        void IPersistStreams.Drop()
+        {
+            this.Drop();
+        }
+
+        public void Purge()
+        {
+            this.TryExecute(() =>
+            {
+                var commitsToDelete = this.QueryCommits().ToList();
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs]: Is this thread safe? What would a commit coming in mid purge process do?

File: NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
+﻿function LogicalKey_Constraint() {
```
Comment (PeterStephenson) [inline NEventstore.Persistence.DocumentDB/Constraints/Logical_Key.js]: An extra thought, we need to ensure that this trigger approach is thread safe too?

**Comment by PeterStephenson** [comment]
Looks good now :) If you tidy your commits a bit I'll merge it :)

**Comment by YannickMeeus** [comment]
Commits tidied up, mergymergy

File: NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs (hunk 9 above)
Comment (YannickMeeus) [reply NEventstore.Persistence.DocumentDB/DocumentPersistenceEngine.cs]: Realised this is still outstanding, don't merge just yet.

**Comment by YannickMeeus** [comment]
@PeterStephenson Mergy Mergy

**Comment by YannickMeeus** [comment]
@PeterStephenson Hola, I'm assuming ASOS has no further need for this repo, so shall I just close this PR and you lot can clean up the repo itself?
